### ISSUE TITLE
Add module-info.java

### ISSUE THREAD

**Issue by hrchu** [google/guava#2970]
So that projects depend on this can be published to a public artifact repository.
Note that this is not breaking backward compatibility. All codes except this file can be still compiled in Java 6.

**Comment by jbduncan** [comment]
Guava has an `Automatic-Module-Name` in its `MANIFEST.MF` now, so I believe this is not quite as important as it may seem. But if I'm mistaken, then I'd be more than happy to be proven so.

(BTW, I think I might have misunderstood something, because `module-info.java` is Java 9 specific, right? Would a Java 8 compiler (which I believe is what is used to compile guava-jre) or an Android compiler (for guava-android) happily process it or otherwise ignore it?)

**Comment by cpovirk** [comment]
Our current thinking is that we'll look into this next quarter. We have seen some problems from `module-info` files in third-party jars, since they're Java 9 `.class` files and not everyone uses Java 9 yet. So, if we can get away with `Automatic-Module-Name`, we'll continue to do so, possibly even beyond next quarter. But if there are cases in which `Automatic-Module-Name` isn't good enough, that would be good to know.

**Comment by hrchu** [comment]
It is possible to only compile ```modile-info.java```, so the jar file is still compatible to users who uses earlier Java version. 

There is a maven example here:
https://github.com/twonote/radosgw-admin4j/blob/java9/pom.xml#L127

**Comment by cpovirk** [comment]
Right, thanks. We've seen problems even when the main `.class` files are compiled for an older version but the `module-info.class` is present. As I understand it, various tools try to process all `.class` files, and they need to be updated to ignore `module-info.class`.

**Comment by hrchu** [comment]
@cpovirk could you tell me more about this problem?

**Comment by cpovirk** [comment]
I wasn't personally involved in fixing the problems, but the basic idea seems to be that people scan the whole classpath (using something like [`ClassPath.getTopLevelClasses`](http://google.github.io/guava/releases/snapshot-jre/api/docs/com/google/common/reflect/ClassPath.html#getTopLevelClasses--) -- which might be an example of something that needs updated to ignore `module-info.class` :)) and then try to examine/load the classes with a tool that understands only, say, Java 8.

**Comment by orionll** [comment]
It's worth mentioning that if we add `module-info.java`, all packages will not be open anymore.

**Comment by jbduncan** [comment]
@orionll Am I right to think/remember that in the JPMS, open packages are packages whose internal classes can be inspected with reflection?

**Comment by orionll** [comment]
@jbduncat Yes, exactly. And also private members of public classes.

**Comment by jbduncan** [comment]
@orionll Cool, thanks for confirming things for me. :)

I personally wonder how important it would be for Guava's packages to be open when used on the module path. I struggle to imagine that reflectively calling Guava's internals is a common thing to do, especially considering Guava's (IMO) pretty durn good API. :thinking:

**Comment by HoldYourWaffle** [comment]
Are there any reasons for it not being open? Even if it's uncommon it might still be done by some people.

**Comment by jbduncan** [comment]
@HoldYourWaffle I think the main reason is it prevents people from using reflection to depend on internals which may change or disappear in future releases of Guava without warning.

**Comment by jbduncan** [comment]
...which by my understanding makes things easier for everyone in the long-run.

**Comment by jbduncan** [comment]
The only reason I can currently think of to have Guava's packages open in the `module-info.java` is if frameworks like Spring need to reflectively access its internals to do important stuff, but I don't know how important or common that is.

**Comment by orionll** [comment]
All Guava packages should be closed because as @jbduncan said the dependence on class internals is a bad practice. If someone **really** wants to access the internals, they can use `--add-opens` command line option which forces the specified package to be open.

**Comment by HoldYourWaffle** [comment]
Good point I forgot about `--add-opens`. Imho you should always be able to hack into internals (maybe you want to do something the developers didn't think of but it's too 'personal' that it's not worthy of a PR), with the risk of it breaking in new versions. `--add-opens` allows for this so it'd indeed be better to close the guava packages.

**Comment by SamCarlberg** [comment]
> Guava has an Automatic-Module-Name in its MANIFEST.MF now, so I believe this is not quite as important as it may seem. But if I'm mistaken, then I'd be more than happy to be proven wrong.

It does mean that `jlink` will not work, since the tool requires a module name to be specified in the `module-info.java` file; automatic module names will not be accepted.

**Comment by jbduncan** [comment]
@hannes-transentials I think it's most likely that Google have not migrated to JDK 11 and adopted modules yet simply because their internal codebase is so mind-bogglingly humongous. ;)

I say this because I remember reading somewhere (or I inferred) that they use Guava or an superset internally, and I also remember they announced a few years ago that they'd finally migrated to JDK 8 after a lot of effort. So I'm sure that they'll announce support for JDK 11 or a later LTS version (and, by extension, modules) when they fully migrate away from Java 8 and when they feel that most of us non-Googlers have moved away from Java 8 too. (I know that _my_ company hasn't done so yet simply because Java 9 was such a freaking big, backwards-incompatible change!)

**Comment by orionll** [comment]
It's worth mentioning that adding `module-info.java` can break some existing tools. For example, I know that IDEA's Osmorc plugin does not work when both `module-info.java` exists and the library is an OSGi bundle (Guava is). So, while the tools are not ready yet, I would abstain from adding `module-info.java` to Guava.

**Comment by hannes-transentials** [comment]
So I either do without modularized applications or stay away from Guava (and many other popular applications)? I somehow hoped that there was some middle ground.

**Comment by talios** [comment]
@hannes-transentials You _could_ make use of something like https://github.com/moditect/moditect to adapt guava and add a module-info.java/.class at your applications side of things as a transitionary work around.

If module-info.java was to be added to guava, hopefully it'd be done so as a modular jar so we don't break java 8< versions.

**Comment by seinecle** [comment]
So Guava is still not natively JPMS compatible in May 2020?

**Comment by cowwoc** [comment]
No. And it isn't hard to add support without breaking pre-Java9 support. As @overheadhunter mentioned, all they need to do is add `META-INF/versions/9/module-info.class` (multi-release JAR) and it would be compatible with both versions.

The easiest way to provide Java9 support is to release a separate artifact for it but if they don't want to maintain multiple artifacts then the multi-release JAR approach would work as well (it's just harder to build).

**Comment by orionll** [comment]
> So Guava is still not natively JPMS compatible in May 2020?

Yes, it's not 100% compatible, but it has an `Automatic-Module-Name` attribute which is better than nothing.

> all they need to do is add `META-INF/versions/9/module-info.class`

This is very simple on first sight, but when you actually start implementing this, you will be surprised by the number of unexpected issues you will need to solve:

- How to compile `module-info.java` to `module-info.class`? You will need Java 9 for this, and the rest of the classes will be compiled with `--release 8`. This is achievable but this is not trivial to implement in the build system.
- What the module declaration should actually contain? It should probably **export** all packages, but should it **open** all packages? Currently, guava is an automatic module, so it exports and opens everything. If we gonna change this, how much user code will be broken? For sure, some users are using reflection on Guava classes, so they will be upset when their code stop working.

**Comment by cowwoc** [comment]
@orionll

1. I recently did this for Hikari (https://github.com/brettwooldridge/HikariCP/pull/1573) and it turns out that maven-compiler-plugin recently added support that makes this even easier (https://github.com/metlos/multi-release-jar-maven-plugin/issues/3#issuecomment-403975603)

2. Start with `exports` most/all packages. If people need `open` let them request it. The only thing we technically promised end-users are public methods. If they are using reflection to access private members then I think it is reasonable to chance breaking their code.

**Comment by cpovirk** [comment]
Thanks to everyone who's been contributing details here.

I thought I had some memory that @jodastephen used to recommend Automatic-Module-Name over a full `module-info` (though I figured things might have changed since then). But in fact I haven't found *any* such recommendation, whether in what he said [on our issue tracker](https://github.com/google/guava/pull/2846#issuecomment-313422497) or [on his blog](https://blog.joda.org/2017/05/java-se-9-jpms-automatic-modules.html).

The closest thing I see to a recommendation against is that he recommends waiting until...

> all of your runtime dependencies have been modularised (either as a full module or with a MANIFEST.MF entry)

We should figure out if our annotation-only dependencies matter, as I don't know that they've all been modularized (though we could probably push for it, and we control some of them ourselves).

(I could see starting with either `exports` (which makes the most sense) or `open` (which seems less likely to immediately break people), possibly moving from one to the other over time.)

**Comment by jodastephen** [comment]
Personally I don't think there is much to gain from adding `module-info`. With my own projects I had to release a parallel jar file without it because it broke downstream users. The time to add `module-info.java` is when Guava moves to Java 11 or 17 as a baseline.

**Comment by overheadhunter** [comment]
> because it broke downstream users

@jodastephen even with MR-Jars? I made the experience that it even works with old Android 4.x apps.

**Comment by cowwoc** [comment]
@jodastephen To @overheadhunter's point, a recent change to Maven makes it very easy to build multi-release JARs. Check out https://issues.apache.org/jira/browse/MCOMPILER-417 and https://github.com/apache/maven-compiler-plugin/blob/master/src/it/multirelease-patterns/singleproject-runtime/pom.xml#L102-L108

**Comment by orionll** [comment]
@jodastephen How exactly can `module-info.class` break the users? Do you have examples?

**Comment by cpovirk** [comment]
I know we have seen this internally: Some tools assume that they can read every `class` file in the classpath. If those tools are written for Java 8, they can have trouble with `module-info`. (Other people may have seen other problems.)

Such tools can and should be fixed. As a practical matter, though, adding a `module-info` to Java can break things, so we end up trying to weigh the costs and benefits.

**Comment by cowwoc** [comment]
Hey Chris. I'm not sure that `module-info.class` would actually be the classpath in this context. If we publish a multi-release JAR and the user is running JDK 8 (because the aforementioned tools break in newer releases) then `module-info.class` shouldn't show up at all.

It would help to look at a concrete testcase. I suspect we'll be okay.

**Comment by jodastephen** [comment]
Tools have been known to scan all class files in each jar file found on the classpath. This is hardly surprising, as the whole multi-release jar file concept is new, why would an old tool know about it.

@cpovirk, weighing the costs and benefits is easy. There are [no benefits to library projects](https://blog.joda.org/2018/03/jpms-negative-benefits.html) like Guava, only costs. The situation might change when Guava has Java 11 or 17 as a baseline. Adding an Automatic-Module-Name is quite sufficient for those who are desperate to use modules.

**Comment by overheadhunter** [comment]
> Some tools assume that they can read every `class` file in the classpath. If those tools are written for Java 8, they can have trouble with `module-info`.

@cpovirk In case of multi-release-jars, the `module-info.class` would be stored at `META-INF/version/{n}/module-info.class` (`{n}` being 9 or 11 or whatever).

Even with Java ≤ 8 it would violate the jar file specification, if tools looked for classes inside of `META-INF`, so regardless of JPMS or newer Java versions, it is legitimate to file a bug, if they do so.

> There are [no benefits to library projects](https://blog.joda.org/2018/03/jpms-negative-benefits.html) like Guava

@jodastephen If you came to this conclusion for your project, fair enough. But could you please not express personal opinions as if they were _facts_?

Especially libraries should carefully consider if they can support JPMS, as they block downstream adaption. The less transitive dependencies there are, the easier a module-info.java can be implemented. Downstream projects can not start adapting JPMS before all their dependencies do so. Automatic module names are just a first step, but don't facilitate all possibilities.

Not only for API design but also from a security perspective, properly restricting access to non-public APIs should be reason enough for project leads to deal with JPMS.

**Comment by cpovirk** [comment]
If people have horror stories about, say, other libraries that have used non-public Guava APIs and thus broken their users, that would help motivate this. (But of course there's a limit to what modules can prevent until they're universally used -- and even then, people could hack around them.) I think we've heard of a library that did something along those lines with gRPC or protobuf or something in the past?

Based on what we've heard, though:

- We know that adopting modules has hurt people whose tools have one kind of bug.
- It's unclear whether adopting modules will help people whose dependencies have another kind of bug.

The more specifics that people are able to share, the better. We have a long track record of doing things that "should be safe" that cause users trouble that we ought to have foreseen :(

**Comment by jodastephen** [comment]
I can't stress enough the aspect that modules do not provide benefits to library authors.

Is adding `module-info` low effort? No. It quadruples the potential testing scenarios, makes the build much more complex (when based on Java 8) and you are quite likely to do it wrong first time and spend months fixing edge cases. Tools like Maven and Javadoc still have major issues with modular projects.

Does adding `module-info` stop users accessing an "internal" package? No. The jar file can be quite legitimately be used on the classpath instead of the module path, and no protection will occur. Users will still access "internal" packages. Giving yourself a false sense of security isn't a benefit.

Does it break downstream users? Yes, many and various reports of this. Sure the tools concerned might be buggy, but so what. Those users will still [complain](https://github.com/JodaOrg/joda-convert/issues/24).

Who does modules benefit? Very few. They are of some minor benefit in a closed world, where everything uses the module system, everything is 100% correct (no mistakes in any module-info file, no split packages, no circular dependencies etc.) The best way to adopt modules would be in a large company that isolates developers from Maven Central and fully vets all open source. Such a centralised enterprise team could download each external jar, add a module-info, check split packages etc and ensure that only the internal modularised version is used within the enterprise. A large cost and probably impractical in reality, as most enterprise software isn't written well enough wrt split packages and circular dependencies to be able to use modules anyway!

I get that people want to use shiny new things, but sometimes shiny and new is not the same as good and useful.

**Comment by cowwoc** [comment]
@cpovirk 

I don't personally care whether people can hack around JPMS boundaries. Publishing `module-info` helps hiding private APIs from the vast majority of users thanks to great IDE support and for the rest of the world the published Javadoc (which makes use of `module-info` itself) makes it quite clear that what they are doing is not supported. Let's optimize for the happy path, which is growing over time.

JPMS reminds me of Generics. Yes, it has its flaws but it will only get better over time. Do you want to be that guy who keeps on exporting raw types for years after Java 5 came out? Yes, people can still use your library but increasingly it becomes a pain in the behind to do so.

So again, let's talk about concrete issues. You keep on mentioning that some people are impacted by the use of broken tools. Can you provide the names and numbers of said tools? Can we count how many people are impacted by this? If there are issues, let's fix them and move forward.

**Comment by overheadhunter** [comment]
It is correct that split packages are a pain in the a** when using the module system. But is it right to demonize the module system or should we rather think about split packages? That said, Guava uses its own namespace so it will only collide with other 3rd parties that claim `com.google.common`, too. To state the obvious: This isn't Guava's problem.

> The jar file can be quite legitimately be used on the classpath instead of the module path, and no protection will occur.

_If_ the library is on the classpath, that is. Would this be a regression? No.

Btw you just made a great point for adding a `module-info`: Users are _not_ forced to use the module system. If you need to stick to the classpath, a modular dependency will do no harm. Just put it on the classpath and you're fine. As I said before, we have a mr-modular-jar mixed with a lot of "normal" dependencies inside an Android 4.x app. The devs don't even notice the difference.

**Comment by johnmanko** [comment]
```
        <dependency>
            <groupId>com.google.code.gson</groupId>
            <artifactId>gson</artifactId>
            <version>2.8.6</version>
        </dependency>
        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>29.0-jre</version>
        </dependency>
        <dependency>
            <groupId>javax</groupId>
            <artifactId>javaee-api</artifactId>
            <version>8.0</version>            
            <scope>provided</scope>
        </dependency>
```


```
module MyModule {
    requires com.google.gson;
    requires com.google.common;
    requires javaee.api;
    requires jersey.guava;
    exports com.my.base;
}
```


`[ERROR] module jersey.guava reads package javax.transaction.xa from both javaee.api and java.transaction.xa`

**Comment by ogregoire** [comment]
Well, thank you GitHub, I searched for the same issue before creating this one. At least Google helped, only after I created this issue. It can be closed as a duplicate of #378.

**Comment by cpovirk** [comment]
A recent experience in JPMS adoption: https://github.com/ben-manes/caffeine/issues/535 -> https://github.com/ben-manes/caffeine/pull/540

**Comment by rbygrave** [comment]
> I can't stress enough the aspect that modules do not provide benefits to library authors.

Well, having added module-info to a library that is together made up of 12 odd modules (plus another 8 modules that are upstream dependencies), for me I would say that module-info forced me as a library author to be more explicit in terms of "public api" (exports) and "friend api" (exports to) and overall I'd say it was a pretty significant improvement for the longer term and a really good push to tighten up and tidy up some areas.  I actually think the module-info provides a really nice and explicit summary (requires, provides, uses, exports) of how things tie together. 

If a library is just a single artifact / module sure, the benefit is a lot smaller but then module-info is also generally super easy in that case.

> Who does modules benefit?

For me, I think it's a longer term thing. We get to be more explicit with "public api" which helps libraries longer term. Using module-path has got a LOT better in tooling (honestly, there was tooling pain but to me that has got a lot better).

My opinion on the benfits of java modules @jodastephen

**Comment by jensli** [comment]
Ping! Making Guava into an explicit module is important! Without a module-info Guava blocks the use of jlink.

**Comment by cowwoc** [comment]
I gave up on Google-authored libraries when it comes to Java9+ support. I hope they prove me wrong.

In the meantime, the following tool works great: https://github.com/moditect/moditect

**Comment by cpovirk** [comment]
It's been long enough that I wouldn't worry about lack of tool support anymore. (While I wouldn't be surprised if _some_ tool still had a problem with modules, I wouldn't expect that to be common, nor would I expect for another year's wait to help.)

So it seems likely that users would be better off if we were a proper module. The next trick would be to actually make that happen:
- The instructions in https://github.com/google/guava/issues/2970#issuecomment-653282535 look like a big piece of the Maven puzzle. My only worry there is that we'll need similar changes for Javadoc generation and perhaps other plugins. I'm sure it's all surmountable; I just would expect this to eat someone's day.
- I suspect that we would end up breaking all links to our snapshot Javadoc. Ideally we'd set up redirects for the new modular directory structure.
- I believe we have various dependencies that are not modularized. I think at least some of them are our fault (or the fault of Google more broadly), but we'd need to get them modularized or, given that most contain annotations only, figure out whether we could get away with "ignoring" them, similar to how there are [consequences to using `provided`/`optional` for them](https://github.com/google/guava/wiki/UseGuavaInYourBuild#what-about-guavas-own-dependencies). I recently heard a claim that we probably could _not_ get away with that :( It's possible that some of the dependencies could be eliminated (e.g., maybe by using custom `@J2ObjCIncompatible` annotations instead of the official one?), including, of course, eventually [migrating off the JSR-305 annotations](https://github.com/google/guava/issues/2960). (I wonder if our dependencies on the annotation artifacts would become "stronger" if we properly modularized Guava? If so, that's one possible downside to doing so.)

It seems likely that that's enough obstacles that we _still_ aren't going to make that happen anytime soon. The point at which it's worth having a deeper look is probably when the JSR-305 dep is finally gone.

**Comment by bowbahdoe** [comment]
For anyone who is interested in using guava with the module system/java9+ support (which looks like @charneykaye , @jensli , @cowwoc, @johnmanko , @seinecle, and probably someone reading this issue) I [made a script](github.com/bowbahdoe/guava-generator) which shades the source code and makes a series of modular jars

github.com/bowbahdoe/guava

Version 0.0.3 is taken from the current latest release of guava and i'll keep it up to date as I can. That may or may not serve your purposes, at least until all the problems here are worked out. You can reach out to me directly or on any forum if you want (but lets not clutter this issue)

But, relevant to this issue (and I think adding a module-info here is pretty important, downstream libraries need the "base" to be modular before they make the same leap) I discovered some things that are blockers in addition to the JSR-305 thing, as well as general "future java issues".

* `LittleEndianByteArray`, `AbstractFuture`, `UnsignedBytes`, and `Striped64` all depend on `sun.misc.Unsafe`. This means that their modules would need to require `jdk.unsupported` in their module-infos which is unfortunate. The good news is that all of their usages of `Unsafe` have supported alternatives in JDK 9+. (All `VarHandle` best I can tell)
* The code to load `sun.misc.Unsafe` in all of those cases uses the security manager. While "use unsafe if we can, fallback if we can't" works today, because the security manager is going to be taken out back and 🔫  this code will become totally incompatible with future java releases. There is also one stray reference to `AccessControlException` in `Types` which can probably be replaced with `Exception` since it is just wanting to "setAccessible(true) if allowed"
* `Throwables` uses `sun.misc.JavaLangAccess`. I think this usage might be "safe" to keep in since the fallback doesn't imply a performance dip, unlike the `Unsafe` usages.
* `FileBackedOutputStream` has behavior that relies on finalization. Finalization is set to be removed from the JDK in a future release, so it might be advisable to provide the same behavior, but with a `Cleaner` for JDK 9+
* There might be no way to make `FinalizableReferenceQueue` continue to work going forward, but maybe it can also have a `Cleaner` based alternative? (My read of that class and associated ones is that they are functionally equivalent to `Cleaner`, but I might be wrong?)
* Usages of `java.util.logging.Logger` aren't problematic, but they are made somewhat "in the corners" of the library. It is unfortunate to have to include the whole logging module when its not really used for much. Because `System.Logger` is in `java.base` and delegates to  `java.util.logging.Logger` by default, it might be possible to backwards compatibly move to that for Java 9+ consumers.

So it seems like no matter what, even ignoring module-info, guava is going to have to provide a multi-release jar. The removal of the security manager will make current guava builds totally unusable eventually and there is no way to keep supporting java 8 without using it.

If anyone is looking for work outside of what @cpovirk is doing w.r.t. the JSR-305 dep. and (I assume) `jspecify`, doing the build plumbing for a multi-release jar feels valuable.

**Comment by cpovirk** [comment]
Thanks for all the investigation.

Random responses:

- `sun.misc.JavaLangAccess` is also used to improve performance. That said, as the deprecations in the class suggest, the `JavaLangAccess` APIs no longer exist as of Java 9, so no one on 9+ is benefitting from it. (This, along with other things that you mention, is another thing for us to think about when we someday do https://github.com/google/guava/issues/6614.)
- cross-reference to another recent discussion that touched on `System.Logger`: https://github.com/google/guava/issues/6566#issuecomment-1599693421
- Another class that uses finalization is `ClosingFuture`. I believe that both `ClosingFuture` and `FileBackedOutputStream` should work OK without finalization as long as people are careful to close them. But of course the finalizers are there exactly to help when people aren't careful enough :)
   And I just found a bug filed by @cgdecker back in 2012 about removing `resetOnFinalize`... :)
   > There are only 2 usages of the feature and both could be changed to not use it. This would eliminate the only occurrence of a finalizer in our libraries.

  Oh, well.
- The `PatternCompiler` functionality to load other implementations is available only inside Google. It's serving one specific use case that we've been working to eliminate, and when the functionality was available outside Google, it caused some problem or other, the details of which I'd forgotten but which appear to be related to... the module system! https://github.com/google/guava/issues/3147
- I wonder if we've had a focused discussion of multi-release jars at some point. I did find a little talk of them in https://github.com/google/guava/issues/6549.

**Comment by jbduncan** [comment]
@bowbahdoe Wow, excellent work! I now see how sorely mistaken I was in my [2017 message](https://github.com/google/guava/issues/2970#issuecomment-337504948) when I asked if `Automatic-Module-Name` was sufficient. It most certainly is not. Well done discovering all this! 👏

**Comment by sgammon** [comment]
Has this seen any motion? We want to use `jlink`, but we can't because Guava is one of the last libraries on the list to ship a `module-info`. JDK 22 is coming this month, so JDK 8 or JDK 11 support as a new minimum would probably be okay for many users.

Guava is also one of the hardest modules to patch, replace, override, or remove entirely

**Comment by bowbahdoe** [comment]
@sgammon this won't help if guava is a transitive dependency, but for what its worth I am still maintaining that modularized mechanical fork. I haven't had time to update to the latest release, but i'll get to that eventually.

**Comment by sgammon** [comment]
@cpovirk I happen to be familiar with [J2CL](https://github.com/google/j2cl/issues/200) and [Bazel](https://github.com/sgammon/rules_graalvm). Let me see if I can isolate a minimal suite of changes. @bowbahdoe Thank you for this link, too, I will see if Moditect can make things simpler.

**Comment by sgammon** [comment]
Hey everybody,

I've gotten started on the work needed to land this. It may take awhile but I think a PR could be accomplished soon, so long as it uses some pre-packaged JARs while we wait for JPMS-enabled releases upstream. If you need Guava to support JPMS, now would be a good time to say so, so that Google knows this is worth the effort (respectfully, of course; don't hound the maintainers or anything).

Guava is #1 in the Core Utilities category on Maven Central, and #4 in terms of overall rank (globally) for all artifacts. Fully modular Java builds require a `--modulepath` and in some cases ban or frustrate the use of Automatic Modules. So, it stands to reason that fully modular Java is out of reach for _most_ apps and libraries, if only because Guava is likely to appear somewhere in the transitive dependency graph.

JPMS is great because it offers faster compile times, better security/isolation in some ways, and so on. It is frustrating in other ways. But certainly it is not going away. Hopefully Guava supporting JPMS can make modular Java a possibility for all of us downstream. 👍

**Comment by johnmanko** [comment]
> If you need Guava to support JPMS, now would be a good time to say so

You got my vote.

**Comment by sgammon** [comment]
Hello everybody!

Firstly, wow, thank you for the support on that comment! As you've probably seen from this issue, there has been a lot of motion on this from Google, and they deserve a big thank you for this.

I'm commenting here now because **Guava with JPMS is ready to try**! It's early, but the tests all pass and there are samples building successfully.

### Current state of things

- All upstream PRs are merged and on their way to release. This includes [Checker Framework](https://github.com/typetools/checker-framework/pull/6326), [J2ObjC](https://github.com/google/j2objc/pull/2302) and [Error Prone](https://github.com/google/error-prone/pull/4311).
- Guava has a [draft PR](https://github.com/google/guava/pull/7094) for JPMS; it will need review and some cleanup
- All of these changes remain compatible with JDK 8, they are issued as multi-release JARs
- There is now a custom Maven repository, called the [JPMS Attic](https://jpms.pkg.st), that you can use to try all this safely

### Library versions

The JPMS Attic repo has all of the above libraries, i.e. everything transitively needed to fully build Guava (+ potentially your library or app) via modular Java. The repository contents are [fully on GitHub](https://github.com/javamodules/attic) for easy inspection. They don't _build_ on GitHub yet (I'd like to do this so there is clear provenance), but they will shortly.

**Until official Maven releases, these libraries have versions suffixed with `-jpms`, in order to avoid conflicts with Maven Central.** Here's the table of artifacts available now:

| Coordinate                                      | Version                 |
| ----------------------------------------------- | ----------------------- |
| `com.google.errorprone:error_prone_annotations` | `2.25.0-jpms`           |
| `com.google.guava:guava`                        | `33.0.0-jre-jpms`       |
| `com.google.j2objc:j2objc-annotations`          | `3.0.0-jpms`            |
| `org.checkerframework:checker-qual`             | `3.43.0-SNAPSHOT`       |

#### Dependency Verification

Obviously, these artifacts are not yet signed with Google's GPG keys. In the meantime I have signed them with mine. This matches commit signatures on GitHub for the changes applied. Key fingerprint is:

```
A166AC62F6DE8C63732277731D5903AF4582BCF4
```

<details>
<summary>Artifact digests</summary>
<pre>
c2681472076c8da3883d5d94fb15b362acc7920e7885d53a652133450c6aa8e0  ./org/checkerframework/checker-qual/maven-metadata.xml
325419bd2c7b84445b6fafc136e9d59421b633f31022cabc4532c68f7cac8a8f  ./org/checkerframework/checker-qual/maven-metadata.xml.md5
2988c067d6d46dd5e65558d6d32172fd676cbcd6f95d2e9ac9e6c3a639f6e1f0  ./org/checkerframework/checker-qual/maven-metadata.xml.sha1
efcf78c2e21b6062e2b8113fcff6a5b7dd36ec555c82f049a7d4eed3338494c4  ./org/checkerframework/checker-qual/3.43.0-SNAPSHOT/checker-qual-3.43.0-SNAPSHOT-sources.jar
dcb92f0425cbd5fa7d9b64b46ff484d7cb9f0ba8348277501ff1f902947f276c  ./org/checkerframework/checker-qual/3.43.0-SNAPSHOT/checker-qual-3.43.0-SNAPSHOT-javadoc.jar
5ac2f9186d54139e804fc88f9cdf17034c19c499fd49ecb790bc417c3c47ad71  ./org/checkerframework/checker-qual/3.43.0-SNAPSHOT/checker-qual-3.43.0-SNAPSHOT.pom
c5ae827b30115e8eb1251c4371a9b2f2c6892be3ded43d736c86b5cdc5d66edf  ./org/checkerframework/checker-qual/3.43.0-SNAPSHOT/checker-qual-3.43.0-SNAPSHOT.module
a8224eb81224e809b741a3cecadc65be5bb3ea29964779fb120476311086f538  ./org/checkerframework/checker-qual/3.43.0-SNAPSHOT/maven-metadata-local.xml
f9a749d838626b2ba60449e267c31fb4257405ad0b2ca03b77d4dce9a7a8d700  ./org/checkerframework/checker-qual/3.43.0-SNAPSHOT/checker-qual-3.43.0-SNAPSHOT.jar
937e488f2e18cebe6f57d22b188ebd2349da43aeef4f51ab077b9abfd2dd53b4  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.081842-1.pom.sha1
62a5ecebfa8bf55a440a2d0ca00c149e380c950ee05db49554be8c64963b613d  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.083856-2.pom.md5
27c4e3990b03b2e8ec47eea9182fea728b0fe934fd33eb89db8fecf32c50e769  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.083856-2.pom.sha1
7486808818886da57969387f3700f8a54b20987d53a4a39c6fd72963a4c06634  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.083856-2.jar.md5
8f1e709404797dabeeb769c0292ad45dcb438eae32885f2c5df0eb217421470d  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/maven-metadata.xml
cb2f49317800e50fa870eb1e7aeaf0bfd0b3a18563af02e910d81779a79670fc  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.081842-1.pom.md5
878c9045e756ff4628c21f78201ac5bfc2b9d2373c91dc6bf79af5add4d4388b  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.081842-1.jar
e0ef91eb0d6f3e3a6d072e7333ce5d56f5c256548fd0845f10117364b42effbc  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.083856-2.jar
32ae05718e1fc12ea801b164248a8ff05aa9b766412fff48142cdd627007274f  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.081842-1.jar.sha1
c1552c96a30d234b219fbe778fa1b29f43d91e0647ba19a9271e8e11ae80b3b8  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.081842-1.jar.md5
36192a54b3ed2d19b4be54700ae9c788f521c37da62390d32ecc9f1584afdd0a  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.081842-1.pom
8a9e74949dee9508fd433bd0915ddef64eef2b36125b5a85b066780da64b2736  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.083856-2.pom
ed8bd810604ff2c1853fc499e8a403ea76d2de6e07763a198296caafe131a081  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/maven-metadata.xml.md5
08bea9d4891ad76bff7a36d7a475ecb95d3b3f89d5a8071cd50f4f4d4cdedfe2  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/maven-metadata.xml.sha1
94d20a8461990e06f5ebb072b6a98ac51c5ff7cc657944dfd7b3b92a8adb12da  ./com/google/guava/guava/1.0-HEAD-jre-SNAPSHOT/guava-1.0-HEAD-jre-20240310.083856-2.jar.sha1
760f8a465b22877547761227fa6fe9e77b68488841beb76283285b8775f862ed  ./com/google/guava/guava/maven-metadata.xml
8b267cd72e328f211c17a38890516ed2203ae877d905e9cdd15a95c632d818a1  ./com/google/guava/guava/33.0.0-jre-jpms/guava-33.0.0-jre-jpms.jar.md5
920e137661a12a7ccb32933568a34e6a05be01d0de87c32ff73749b9f8c506d4  ./com/google/guava/guava/33.0.0-jre-jpms/guava-33.0.0-jre-jpms.pom.md5
84626ef4bfbef95e2f9d7d0983a0e8829376697a7d07d6c3de328d445f9b88b8  ./com/google/guava/guava/33.0.0-jre-jpms/guava-33.0.0-jre-jpms.jar.sha1
3ed27fa5bdc2589d1d1227c4c99c1db1c0117c7f30bd6acebddb1ddd58f960fd  ./com/google/guava/guava/33.0.0-jre-jpms/guava-33.0.0-jre-jpms.jar
a7f41c1cfea3fd7eb0fc305a155b7d4931fc692a5c0a077d202d92afed14d82c  ./com/google/guava/guava/33.0.0-jre-jpms/guava-33.0.0-jre-jpms.pom.sha1
9fe95275923730b89704c2fd6d85fd7a8880e688716a558aee61f8f5fb60895f  ./com/google/guava/guava/33.0.0-jre-jpms/guava-33.0.0-jre-jpms.pom
3c0b16b1c401730cd57519f4a9cdb459fab977634eb05396dec238308e06189d  ./com/google/guava/guava/maven-metadata.xml.md5
4c62d866a06ec05f7d7cc1299534229863f9e505afd7e78bd80fa9bdb0482e1e  ./com/google/guava/guava/maven-metadata.xml.sha1
1c65d6a374eb3f684c95f000a40f8f0788af68bdbc447ec0d17e1340b5ffc49a  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081248-1.jar.sha1
e0749d42cc6614ce42e26dc03191699a8334bc402163317618a4983966cd4d04  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081248-1.pom.md5
e0749d42cc6614ce42e26dc03191699a8334bc402163317618a4983966cd4d04  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081646-2.pom.md5
e0749d42cc6614ce42e26dc03191699a8334bc402163317618a4983966cd4d04  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.082021-3.pom.md5
27e0c25783858499a345d0c18fcab48321eff522f69771eba639f756dd46a627  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081646-2.pom.sha1
27e0c25783858499a345d0c18fcab48321eff522f69771eba639f756dd46a627  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.082021-3.pom.sha1
ef8f13b9ff45d4c119182e6b73c2a74bc5d6116d9ad7493f5f46c082dcd6b03a  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081646-2.jar.md5
3f7e72680b2259fb6f184419ec2e4f899bc84ce2e1a5efe31ec79cd689ae970b  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.082021-3.jar.md5
18e17b699a5d219f8cd94f96bb292808229e7b78617ef5db8d7c5001982967c1  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081248-1.jar.md5
26f492b9687743f5967163888ca5e541d8786d545ae0927c16fae37844f4d9a6  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/maven-metadata.xml
27b6ead9a23ee464888362377dcfdea1e290f354cc1bc92e4f008baa06fbb26b  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081248-1.pom
27b6ead9a23ee464888362377dcfdea1e290f354cc1bc92e4f008baa06fbb26b  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.082021-3.pom
06f2517b8fafa39fc33ffd58a22dc07d13e86dbc4aeb5ed5143935df8bceed71  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081646-2.jar.sha1
27e0c25783858499a345d0c18fcab48321eff522f69771eba639f756dd46a627  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081248-1.pom.sha1
5010a3822cc27993da6449b6af0965a068d902fbc284ac0591dbe3eacc6cae75  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.082021-3.jar.sha1
27b6ead9a23ee464888362377dcfdea1e290f354cc1bc92e4f008baa06fbb26b  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081646-2.pom
15bfdd35dd2804671b0b862ad2ce63a4b567e49cc251cb5c6e0ddbddad287668  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/maven-metadata.xml.md5
33b2b200ac0465a05da1560a77a27ad204cb7f46a53c098815804c0fdb8fe993  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/maven-metadata.xml.sha1
c8e70ea20643f4fb3a4b739ba843828fe275d3277a32d857271b6364d7d1bc22  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081248-1.jar
dea47c192130c6524a8c3743661f614d6e0597bd3e4085be49a54313864e1d55  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.082021-3.jar
a75dcc3fa96f9c296e399522883b66a12349b718955115d1e457f0f30d0962de  ./com/google/errorprone/error_prone_annotations/1.0-HEAD-SNAPSHOT/error_prone_annotations-1.0-HEAD-20240310.081646-2.jar
511b91d2a05c18a0384491fb27ca6b1b3354e8dd7a5fe42a70d2f8058fb9e4b3  ./com/google/errorprone/error_prone_annotations/maven-metadata.xml
b9f01dc5b94ad3b25c84f82c53949c8e035d54f88fd839130f45507a53e5f6c4  ./com/google/errorprone/error_prone_annotations/maven-metadata.xml.md5
604f844f85afee7607b96a639e4b404f43a8431c964c87ca34c1dfed346fb2a2  ./com/google/errorprone/error_prone_annotations/maven-metadata.xml.sha1
081ed76568a21a1c23a2f70b793cad4003550ee8ece0f89974b7217af44dbf27  ./com/google/errorprone/error_prone_annotations/2.25.0-jpms/error_prone_annotations-2.25.0-jpms.jar.md5
3b0060a8dec47f9a0281d381ef9431b3b2409ab211a93e1e79208f4239feb6ce  ./com/google/errorprone/error_prone_annotations/2.25.0-jpms/error_prone_annotations-2.25.0-jpms.pom.md5
89a49c9861b01fecc123f469c856b4b2e166eecf7a1fdabdbc9c30f06e33a6d4  ./com/google/errorprone/error_prone_annotations/2.25.0-jpms/error_prone_annotations-2.25.0-jpms.jar.sha1
8745e4577e4c23611b021641bbded603bfbac9edc322bec41623f909aefa88a1  ./com/google/errorprone/error_prone_annotations/2.25.0-jpms/error_prone_annotations-2.25.0-jpms.pom
65fbc0149dca836a304bfb74b4c7d113d278ec8e4f83c9687cbeef130836bfa1  ./com/google/errorprone/error_prone_annotations/2.25.0-jpms/error_prone_annotations-2.25.0-jpms.jar
bb48f614d531d53bde372b0d3f6e6bf8b8ec1787b3388f24db64600192b4b923  ./com/google/errorprone/error_prone_annotations/2.25.0-jpms/error_prone_annotations-2.25.0-jpms.pom.sha1
a813595eb6988d03fca84942769cddd7fc9ec324ba298240e5561c6abc31d807  ./com/google/j2objc/j2objc-annotations/maven-metadata.xml
4b157bf15235b1925c3695b1c7804db6dd003a6e09ebd2e5db947e6ac57d6e60  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081510-1.pom.sha1
4b157bf15235b1925c3695b1c7804db6dd003a6e09ebd2e5db947e6ac57d6e60  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081957-2.pom.sha1
54998eccf2d254a6bb22e4708836ef60c5b6d84de57a1dc6fc04f42620af343d  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081510-1.jar.md5
4f39274b2109a8f7454fd465866c0ce77ceef2cd5dac1b23231d6ed05b1761a0  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081510-1.jar
67b549d39a434ef9a9dd115547dc20fe6b55824ad9676670c2390186b887f19f  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081510-1.pom
b7f1422215780d7489899517a15398d7931a5a9dee6d26a483b53ad92aef08bb  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081510-1.pom.md5
0d8103e18a795c1c3c31265d56794e08784a80f8205fc49da86c4c7f60520337  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/maven-metadata.xml
ffad15ad9464b46b53d2d1a59c07319f07c1e69a2cd10f06606a97854cbe9ec9  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081957-2.jar.sha1
67b549d39a434ef9a9dd115547dc20fe6b55824ad9676670c2390186b887f19f  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081957-2.pom
4670e532139162c19780fdfd08d74a3a76938ff9f435e7d16a4344c9fd96412f  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081510-1.jar.sha1
92814066e7a746ef28fb6c4911efcc097bcd8ad1793922de092c62361a8fbce9  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081957-2.jar.md5
ba3dfdf2fb06dcd8d65f97b55509ac6a60e23d1fbe3dd87881803200204712d8  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/maven-metadata.xml.md5
4b541101d7088ccd6bc4cf540d7df52c650db475fd79f3359b779cd6a28691e2  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081957-2.jar
1cf75740c899452f064448638f96a2cf3103eb4e2ddd566736065dad64495bb1  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/maven-metadata.xml.sha1
b7f1422215780d7489899517a15398d7931a5a9dee6d26a483b53ad92aef08bb  ./com/google/j2objc/j2objc-annotations/3.0.0-SNAPSHOT/j2objc-annotations-3.0.0-20240310.081957-2.pom.md5
caf413a656b17e817fb7a6750fa34423a0f9c59a8b630ba3ea065fdb6732e8ba  ./com/google/j2objc/j2objc-annotations/maven-metadata.xml.md5
07c18b57336a6c1e06c4c6810bd92442f5ceafbc7fdc52a964fd352d74a2a3a9  ./com/google/j2objc/j2objc-annotations/maven-metadata.xml.sha1
c9528d7c15ae180a24ff248ed189fd1e7fd12ed84ea7a8a81ef8d23dd1a1919d  ./com/google/j2objc/j2objc-annotations/3.0.0-jpms/j2objc-annotations-3.0.0-jpms.pom.md5
e6f5e5c4471eedba428638b51475c54b789e9c695fd3905a11c5816107bc5fdb  ./com/google/j2objc/j2objc-annotations/3.0.0-jpms/j2objc-annotations-3.0.0-jpms.jar.sha1
29994c04ca63f8f81d26ae36a1f90ff59454238dead14e4d33cc80af8fcc488f  ./com/google/j2objc/j2objc-annotations/3.0.0-jpms/j2objc-annotations-3.0.0-jpms.jar.md5
cd20c532f794f30046a089354727b215788a7077bc3ae34bc6cab31d4ca7de81  ./com/google/j2objc/j2objc-annotations/3.0.0-jpms/j2objc-annotations-3.0.0-jpms.pom
a9575f66820a1d03d78da53e2e60a290eaadbe2543837d09e6a28be18b17847e  ./com/google/j2objc/j2objc-annotations/3.0.0-jpms/j2objc-annotations-3.0.0-jpms.jar
419b97e5e51386e6e301015262795ad212d0d6f698e7a99db35dc5029ddd6433  ./com/google/j2objc/j2objc-annotations/3.0.0-jpms/j2objc-annotations-3.0.0-jpms.pom.sha1

</pre>
</details>

### Trying it out

Here are some inlined instructions for using the [JPMS Attic](https://jpms.pkg.st) repo in your build. If something isn't covered here that you need, check the GitHub repo, file an issue, or let me know here with a tag:

#### Maven

```xml
  <repositories>
    <repository>
      <id>jpms-attic</id>
      <name>JPMS Attic</name>
      <url>https://jpms.pkg.st/repository</url>
    </repository>
  </repositories>
```

#### Gradle

> There is also a [Gradle sample](https://github.com/javamodules/attic/tree/main/samples) used in integration tests.

##### Groovy

```groovy
repositories {
    maven {
        url "https://jpms.pkg.st/repository"
    }
}
```

##### Kotlin

```kotlin
repositories {
    maven {
        url = uri("https://jpms.pkg.st/repository")
    }
}
```

**Comment by sgammon** [comment]
Hey @ben-manes,

Firstly, big fan of your work... Gradle Versions, Caffeine, so many libraries. I built an entire company (partly) on Caffeine. Shameless plug for [Buildless](https://less.build), which soon will be fully free forever for open source.

I hope you will allow me to respond out-of-order here.

> Your PRs are interesting and good work [...] I really like your ideas and direction

Want to say ahead of the rest of this that I will frame this in my mind. You're a giant and that's very kind. Thank you!

> It is often best to think of Google OSS as shared with the community, rather than driven by it, so as to realize it closely follows Google's internal needs rather than external ones

[You could not be more correct and that is a very healthy way to think about it.](https://github.com/firebase/firebase-ios-sdk/issues/6115#issuecomment-1806420731).

> Please be aware of this comment and Guava's general stance on [contributions](https://github.com/google/guava/wiki/HowToContribute#code-contributions). A lot of necessary work, but also racing well ahead of the maintainers and their priorities

I am aware, and I know that JPMS is arguably a new feature, even though it maybe shouldn't be considered as such. I could argue that we're not _that_ ahead of the maintainers here, given that the Guava team has had this under consideration for some time (speaking about just JPMS here for a moment). Your point stands, though, regardless of those arguments, that JPMS is not a priority inside Google, at least for OSS, and @cpovirk has been clear and honest about that, to his great credit.

Luckily, I've contributed to libraries at Google before (though nothing as widely used as Guava), so this is not a surprise at all and I totally understand that Googlers straddle several (sometimes diverging) priorities. The first few of these PRs are already merged on Error Prone, J2ObJC, and on Guava.

> but also quite large with Caliper to JMH migration, Java clean up, and lots of CI & build additions

This was and remains my fear as well, and so I have split off the JDK 21 and CI stuff entirely. Each stands on its own, and the CI stuff is entirely optional. If the Guava team chooses not to merge these, no worries, we can rebase the JPMS pull cleanly and move on.

That said, I think you're right about Caliper vs. JMH. During that work I actually did figure out that the benchmarks were running from Maven, though probably not... reporting anywhere or doing anything. In any case, they should probably be left alone so that internal Google tooling stays working. At the very least this should be another split point in the PR.

The bench changes, thankfully, were made very late in the PR and stand alone, so those can be split away cleanly.

> And certainly don't get demotivated by what I say, just be realistic that your work might be a PoC that guides them to support once that becomes a necessity

This is excellent feedback and I deeply appreciate you taking the time to share it. I don't find it demotivating at all. In fact, I am only further motivated because the feedback that has poured in so far has been great, including yours. Some of it positive, some of it not, but all really really useful and helpful and engaged.

I am consuming these artifacts myself for my own projects from the JPMS attic repo, and they are working fine after a few tweaks, so my code is not blocked by this, and others have an option to try it out early, too. I think, all of the above considered, the Guava project can take their time and we (the group of people who care about this, in this issue) can test, refine, and eventually get a very clear plan going that minimizes breakage downstream.

Anyway, thank you for sharing your thoughts. Honesty is an expensive gift.

**Comment by jbduncan** [comment]
@sgammon I'm in awe of your realistic and professional yet persistent approach to getting JPMS support and a better CI setup into Guava, as I'm acutely aware how Guava is more of a "throw over the fence" project than a community project, so well done and I wish you all the best moving things forward! 👍

**Comment by sgammon** [comment]
@jbduncan Thank you! That's very kind of you, but I'm standing on the shoulders of giants here. This issue was filed nearly 10 years ago. We should make jackets.

**Comment by bowbahdoe** [comment]
@sgammon I'll take one, but maybe just `GUAVA 2970` on the front. I don't actually want anyone to ask me questions

**Comment by cpovirk** [comment]
The toolchain setup of https://github.com/google/guava/pull/7333 may lay the groundwork for continuing to run tests under JDK 8 even after we requires JDK 9 to build (at least if you want `module-info`). It may well be duplicative of existing work in #7094, which I still need to get back to :(

**Comment by sgammon** [comment]
@cpovirk Thank you for the ping; no worries, I'm happy to rebase or construct it again from scratch if needed.

**Comment by sgammon** [comment]
@cpovirk I intend to pick this up again in the next few days, assuming it's still helpful to keep it rebased and functional. It looks like many changes needed for this were applied to `master`, so #7094 should get much cleaner.

**Comment by cpovirk** [comment]
Hi, @sgammon. There is a bunch of good stuff in that PR, including the attestation work (which Google is probably going to want us to do eventually, though it's possible that they'll also provide us with some amount of standard infrastructure to handle some of the work). Given my track record (say, an unreplied-to email from you from early last year, I'm sorry) and the various other things I'm behind on, the best hope for starting to land things is to isolate changes as much as possible. I suspect that `module-info` is still the biggest user request at the moment, so if it's possible to extract the bare minimum for that, then that will improve the odds that I actually carve off time for it (and then hopefully find a time that I'd be able to look at more in the future).

**Comment by sgammon** [comment]
@cpovirk Sounds good and I agree. I want to winnow this down to exactly and only the `module-info.java` stuff. Lots of these ancillary changes have already made it in (for example, the compiler upgrades are all done!) so it should be an easier trip than it was last time. No worries on unreplied-email, I am a terrible offender in that regard :)

**Comment by cpovirk** [comment]
Good question. That will be a question for someone who knows more about modules than I do :) But to recap for such people: `guava-jre`\[*\] still includes code that uses `Unsafe`, but that code will never be run under Java 9+. Depending how much enforcement we/users end up getting at build time vs. how much at runtime, we may or may not be able to get away without listing `jdk.unsupported`.

(Maybe the module system will end up being [another](https://github.com/google/guava/issues/6806#issuecomment-2637862188) reason to consider [a multi-release jar](https://github.com/google/guava/issues/5621), this time even for `guava-jre`?)

Given what I said a few hours ago, I'd be interested in knowing if we could "just" include `jdk.unsupported` to start and then attempt to remove it later. I understand that its presence would naturally be _concerning_ to people, but given that it's just an acknowledgment of a pre-existing reality (and, again, that it's not really used under Java 9+), I could feel OK with that for now—as long as including `jdk.unsupported` doesn't cause some _concrete_ problem?

\[*\] `guava-android`, which can be run under a JRE, is a different story: It would still use `Unsafe`, at least for now.

**Comment by cowwoc** [comment]
I can't think of any problems with you depending on `jdk.unsupported` and removing it later. Go for it.

**Comment by sgammon** [comment]
@cpovirk I've been reworking this PR over the past 48 hours, and it is apparent to me that there are multiple approaches we could take for JPMS compilation steps:

(1) **Modify JPMS-enabled libs only.** Additional compiler executions could be added to JPMS-enabled modules (non-Android `guava`, `failureaccess`, `testlib`), with defaults left alone in the top-level POM and elsewhere.

  - This repeats some configuration for sure, but is probably less invasive for Android and GWT use cases
  - The PR takes this approach now, but I wanted to try it against 2 for comparison

(2) **Modify defaults and then non-JPMS libs.** More libraries are JPMS-enabled in this PR than not (the odd ones out are Android Guava, `guava-gwt`, and `guava-test`). So, maybe we set new defaults and override modules which do not need or cannot support JPMS.

  - Initial investigation reveals that this might not reduce the size of the change by much
  - We would have to get overrides exactly-right in the more exotic leaf libraries
  - There are some complications down this route related to compile source roots -- JPMS wants them, but they preclude use of `<exclude>` etc., as I know you are aware from inline comments ;)

Any advice one way or the other? After investigating 2, I'm inclined to stick with 1, with extensive inline comments justifying why compiler flags are put where they are.

Current draft of these changes is [here](https://github.com/sgammon/guava/pull/43), although it is undergoing cleanup now and will trim down. Case 1 and 2 are both on the order of hundreds of lines, less than 500 for sure. We're talking about, like, 350 vs. 250 or somewhere in there in terms of lines added, mostly of POM XML:

![Image](https://github.com/user-attachments/assets/f18ef897-8acd-4c28-b345-3e94e2717829)

The original PR weighs in much heavier:

![Image](https://github.com/user-attachments/assets/8d2d43ef-cd2a-4e57-a430-a66db68d839d)

So I think we are well on our way to a trim, potentially mergeable change.

**Comment by sgammon** [comment]
Update: You can try the latest JPMS-enabled Guava [here, in the JPMS attic repository](https://github.com/javamodules/attic). Add it to your Gradle or Maven build according to the directions, and then use the JPMS version of Guava with:

```xml
<dependency>
  <groupId>com.google.guava</groupId>
  <artifactId>guava</groupId>
  <version>33.4.0-jre-jpms</version>
</dependency>
```
```kotlin
dependencies {
  implementation("com.google.guava:guava:33.4.0-jre-jpms")
}
```

To force overrides transitively in Gradle:
```gradle
configurations.all {
  resolutionStrategy.eachDependency {
    if (it.requested.group == "com.google.guava") {
      useVersion("33.4.0-jre-jpms")
      because("jpms support")
    }
  }
}
```

Several [projects are passing integration tests, including Caffeine](https://github.com/javamodules/attic/tree/main/tests/integration). I am using it already in downstream projects via Gradle with no complaints.

**Comment by cpovirk** [comment]
I don't want to act like I would have foreseen this, but now that you say it, I find it very believable that (1) would turn out simpler. I skimmed the draft PR, and nothing in it immediately offended my delicate sensibilities, so that also seems like a good sign. I'm all ears if you or others end up with reasons to prefer (2) in the future, but (1) sounds fine to me.

**Comment by sgammon** [comment]
Cool :) once it is cleaned up and my mess of commits are squashed, I will update the attached PR (#7094) and push it again to the JPMS attic for testing.

**Comment by sgammon** [comment]
@cpovirk and others: PR #7094 is ready for review.

**Comment by bowbahdoe** [comment]
@sgammon Unironically can we get the jacket orders now

**Comment by jjohannes** [comment]
Thanks so much for getting this done @sgammon and @cpovirk!
I discovered this small follow up regarding annotation processing when trying it out: https://github.com/google/guava/pull/7732

**Comment by sgammon** [comment]
@bowbahdoe I kind of want to actually make them. I'll look for a low-min jacket provider 😆 Also, I just wanted to extend a thank you to @cpovirk, @bowbahdoe, @cowwoc, @ben-manes, and many others in this thread who were extremely helpful and patient. This PR was easy to write because of all the research that was done ahead of time.

I'm so glad Guava will support modules now. Yay! We did it guys

**Comment by norrisjeremy** [comment]
Folks, the new release seems to have a serious regression in that it doubles the size of the Guava jar file, from approximately 3MB to 6MB.
This seems to be due to the fact that all classes are duplicated under `META-INF/versions/9`.
I'm not sure why it was decided to build a second Java 9 version of all classes, instead of simply putting `module-info.class` under `META-INF/versions/9`.

**Comment by cpovirk** [comment]
Yes, sorry, that was [a misconfiguration](https://github.com/google/guava/issues/6614#issuecomment-2748893334), which obviously I should have tested before publishing the release :( It's mentioned in the current release notes, and I have to have a fixed version published today.

**Comment by sgammon** [comment]
@norrisjeremy This is probably my fault. I encountered this issue with the JAR and OSGi plugin during the JPMS PR, but thought I had solved it.

In any case, extra classes in the versioned JAR root won't preclude anything from working. The JVM will prefer those classes but otherwise they should be identical.

The next release should fix the issue.

@cpovirk sorry about this. I promise I tested as much as I feasibly could locally. I did try to build the module descriptor without sources, but at the time sources were required for other complex reasons (iirc).

**Comment by ben-manes** [comment]
I believe that the testlib's module-info should also require junit, e.g. `requires transitive junit;`. While javac doesn't complain to fail the build, Eclipse's ECJ does produce this build error. Historically ecj was more standard compliant than javac and helped to squash many bugs such as in generics.

> The type junit.framework.TestCase cannot be resolved. It is indirectly referenced from required type com.google.common.testing.AbstractPackageSanityTests

**Comment by cpovirk** [comment]
The PR looked pretty clearly like the result of solving a dozen similar issues, so I think an extra-large jar with some `-Xlint:all` warnings is a pretty impressive outcome for the initial release :) (I, too, had [thought](https://github.com/google/guava/pull/7741) that javac required you pass "real" sources along with the `module-info`.)

For JUnit 4:
- From my testing, it seems to be fine that JUnit 4 has [only an `Automatic-Module-Name`](https://github.com/junit-team/junit4/pull/1571). (That's good, as AFAIK the JUnit 4 folks are still hoping [never to have to publish a release again](https://github.com/junit-team/junit4/pull/1771#issuecomment-1960288311) :))
- I suspect that `requires` would be enough to satisfy the compiler. But `requires transitive` seems [better](https://dev.java/learn/modules/implied-readability/) unless you're the kind of person who doesn't believe in `requires transitive` at all :)
- Once we buy into `requires transitive`, `guava-testlib` should probably be using it for [its dependency on `guava`](https://github.com/google/guava/blob/486d8de7dbb5d6cb02f08dba27f57bad9ed84bca/guava-testlib/src/module-info.java#L18), too, given that `guava-testlib` uses `guava` types in its API. For that matter, _`guava`_ should probably be using `requires transitive` for [its dependency on `com.google.common.util.concurrent.internal`](https://github.com/google/guava/blob/486d8de7dbb5d6cb02f08dba27f57bad9ed84bca/guava/src/module-info.java#L20), given that `guava` [uses types from that module in its API](https://github.com/google/guava/blob/486d8de7dbb5d6cb02f08dba27f57bad9ed84bca/guava/src/com/google/common/util/concurrent/AbstractFutureState.java#L51). Maybe that would eliminate the apparent need for `guava-testlib` to declare [what would then become a redundant dependency on `c.g.c.util.concurrent.internal`](https://github.com/google/guava/blob/486d8de7dbb5d6cb02f08dba27f57bad9ed84bca/guava-testlib/src/module-info.java#L19)? (I don't _think_ that we expose [`java.logging`](https://github.com/google/guava/blob/486d8de7dbb5d6cb02f08dba27f57bad9ed84bca/guava/src/module-info.java#L19) in our API (neither in `guava` nor `guava-testlib`), at least.)

**Comment by cpovirk** [comment]
Check my proposed changes here: https://github.com/google/guava/pull/7748

History shows that a number of you folks know more about all this than we do :)

**Comment by cowwoc** [comment]
@cpovirk I am only concerned about https://github.com/google/guava/pull/7748/files#diff-6eabd8d281c8a36ef65bc60a72f3e4af026d190937553b6b5ac0b9ddb2a6d190R20

By adding `transitive` you are exporting an internal package to the public. That sounds wrong to me.

**Comment by overheadhunter** [comment]
> By adding `transitive` you are exporting an internal package to the public. That sounds wrong to me.

I agree, maybe it is better to relocate `InternalFutureFailureAccess` to a public package? If `com.google.common.util.concurrent.internal` has never been intended to be used by lib consumers, I wouldn't even consider moving it a breaking change.

**Comment by vietj** [comment]
@sgammon thanks for your effort

**Comment by cpovirk** [comment]
Thanks for having a look.

"`internal`" was a poor choice of package name on my part. It _is_ actually meant to be accessible to other users. (And among the reasons that it's a poor name is that our _other_ `internal` subpackage, the one under `common.base`, is _not_ meant to be accessible to other users.) We were trying to steer "normal" users away from it, but we do want code (mainly low-level libraries and frameworks) to be able to use it to maximize performance, like [here in kotlinx.coroutines](https://github.com/Kotlin/kotlinx.coroutines/blob/45893cec51c63490ce294e46ae25cef3e4d625bf/integration/kotlinx-coroutines-guava/src/ListenableFuture.kt#L105). Perhaps other people have wanted to use it but been scared off by the name :(

**Comment by cowwoc** [comment]
Hmm, where *is* this package anyway? :) The closest I got to finding it was https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/Internal.java but this is a class, not a package.

**Comment by cowwoc** [comment]
Nevermind, found it: https://github.com/google/guava/tree/master/futures/failureaccess/src/com/google/common/util/concurrent/internal

And then I found this README.md at the top-level directory:

> The modules under this directory will be released exactly once each. Once that happens, there will be no need to ever update any files here.

Oops... Are we in trouble?

**Comment by cpovirk** [comment]
The package has been around, publicly accessible, and endorsed by us for usage a few years at this point, so I think that's as good as "exported." (That includes being listed in the OSGi `Export-Package` line for the `failureaccess` jar ever since we set up OSGi there, and it includes being in the `exports` section of the Java module that we recently set up.) So I think it's too late to try to rename it, since that would cause the usual havoc of incompatible Guava changes.

As for the multiple releases: The main goal has been to neither add nor remove APIs. I think that 1.0.1 was to add the OSGi information, 1.0.2 was to add an Automatic-Module-Name, and 1.0.3 was to set up a proper module. It would have been idea to do that all in 1.0, but I think that the multiple releases have been only a marginal contributor to the occasional annoyances that people see with `failureaccess`. (But people should correct me if I'm wrong: While I _hope_ that we don't have reason to release a 1.0.4 in the future, I would like to know both the pros and cons that such a release might have.)
